tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -0,0 +1,2918 @@
+  auto tfl_fc_op = cast<TFL::FullyConnectedOp>(op);
+
+  auto output_type =
+      tfl_fc_op.getResult(0).getType().dyn_cast<RankedTensorType>();
+  // Not a ranked tensor output
+  if (!output_type) return failure();
+
+  auto input_type = tfl_fc_op.input().getType().dyn_cast<RankedTensorType>();
+  auto filter_type = tfl_fc_op.filter().getType().dyn_cast<RankedTensorType>();
+  auto bias_type = tfl_fc_op.bias().getType().dyn_cast<RankedTensorType>();
+  if (!input_type || !filter_type) return failure();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool filter_is_qtype =
+      filter_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::QuantizedType>();
+
+  if ((input_is_qtype != filter_is_qtype) ||
+      (input_is_qtype != output_is_qtype)) {
+    op->emitOpError(
+        "ConvertTFLFullyConnectedOp: input/filter/output tensor should "
+        "be all quantized or all native");
+    return failure();
+  }
+
+  Value input_val = tfl_fc_op.input();
+
+  // tfl.fully_connected() can takes various dimension tensor as input
+  // need to reshape it to rank 2 tensor, which tosa.fully_connected only
+  // supports if input tensor is rank 4.  It's not always reshaping to (dim[0] *
+  // dim[1], dim[2] * dim[3]).
+
+  // In some networks it's reshaping to (dim[0], dim[1] * dim[2] * dim[3]) so a
+  // more general way to determine the reshape's shape is by looking at filter's
+  // shape[1].
+  if (input_type.getRank() != 2) {
+    int64_t num_elems = filter_type.getShape()[1];
+    int64_t num_batch = input_type.getNumElements() / num_elems;
+    SmallVector<int64_t, 2> shape_vals({num_batch, num_elems});
+
+    auto reshape_type = RankedTensorType::get(ArrayRef<int64_t>(shape_vals),
+                                              input_type.getElementType());
+    auto reshape_op = rewriter.create<tosa::ReshapeOp>(
+        op->getLoc(), reshape_type, tfl_fc_op.input(),
+        rewriter.getI64ArrayAttr(shape_vals));
+
+    input_val = reshape_op.getResult();
+  }
+
+  Value bias_val;
+  if (!bias_type) {
+    // For some matmuls, the bias may actually be a "UnitType" which has no
+    // value. TOSA requires bias to be an array of output_channel_count values,
+    // so create a constant of the appropriate number and type of zeros.
+    SmallVector<int64_t, 1> bias_shape({filter_type.getShape()[0]});
+    auto bias_type = RankedTensorType::get(ArrayRef<int64_t>(bias_shape),
+                                           input_type.getElementType());
+
+    DenseElementsAttr bias_attr;
+    if (input_type.getElementType().isa<FloatType>()) {
+      SmallVector<float, 2> bias_arr(bias_shape[0]);
+
+      for (int i = 0; i < bias_shape[0]; i++) {
+        bias_arr[i] = 0.0;
+      }
+      bias_attr = DenseElementsAttr::get(bias_type,
+                                         llvm::makeArrayRef<float>(bias_arr));
```

Comment:
I don't think you need makeArrayRef (should implicit cast)

Reply:
Using bias_arr directly generates build error.